src/parquet/file/writer-internal.cc

```diff
@@ -0,0 +1,260 @@
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "parquet/file/writer-internal.h"
+
+#include "parquet/column/writer.h"
+#include "parquet/schema/converter.h"
+#include "parquet/thrift/util.h"
+#include "parquet/util/output.h"
+
+using parquet::schema::GroupNode;
+using parquet::schema::SchemaFlattener;
+
+namespace parquet {
+
+// FIXME: copied from reader-internal.cc
+static constexpr uint8_t PARQUET_MAGIC[4] = {'P', 'A', 'R', '1'};
+
+// ----------------------------------------------------------------------
+// SerializedPageWriter
+
+SerializedPageWriter::SerializedPageWriter(OutputStream* sink,
+        Compression::type codec, format::ColumnChunk* metadata,
+        MemoryAllocator* allocator) : sink_(sink), metadata_(metadata),
+        allocator_(allocator) {
+  compressor_ = Codec::Create(codec);
+  // Currently we directly start with the data page
+  metadata_->meta_data.__set_data_page_offset(sink_->Tell());
+  metadata_->meta_data.__set_codec(ToThrift(codec));
+}
+
+void SerializedPageWriter::Close() {}
+
+void SerializedPageWriter::AddEncoding(Encoding::type encoding) {
+  auto it = std::find(metadata_->meta_data.encodings.begin(),
+      metadata_->meta_data.encodings.end(), ToThrift(encoding));
+  if (it != metadata_->meta_data.encodings.end()) {
+    metadata_->meta_data.encodings.push_back(ToThrift(encoding));
+  }
+}
+
+int64_t SerializedPageWriter::WriteDataPage(int32_t num_rows, int32_t num_values,
+    const std::shared_ptr<Buffer>& definition_levels,
+    Encoding::type definition_level_encoding,
+    const std::shared_ptr<Buffer>& repetition_levels,
+    Encoding::type repetition_level_encoding,
+    const std::shared_ptr<Buffer>& values, Encoding::type encoding) {
+  int64_t uncompressed_size = definition_levels->size() + repetition_levels->size()
+    + values->size();
+
+  // Concatenate data into a single buffer
+  std::shared_ptr<OwnedMutableBuffer> uncompressed_data =
+    std::make_shared<OwnedMutableBuffer>(uncompressed_size, allocator_);
+  uint8_t* uncompressed_ptr = uncompressed_data->mutable_data();
+  memcpy(uncompressed_ptr, repetition_levels->data(), repetition_levels->size());
+  uncompressed_ptr += repetition_levels->size();
+  memcpy(uncompressed_ptr, definition_levels->data(), definition_levels->size());
+  uncompressed_ptr += definition_levels->size();
+  memcpy(uncompressed_ptr, values->data(), values->size());
```

Comment:
Add another TODO to reuse the same buffer for uncompressed page data on its way to being compressed (to avoid alloc-dealloc issues)